src/Renci.SshNet/Security/KeyHostAlgorithm.cs

```diff
@@ -101,7 +102,11 @@ private set
                     _keys = new List<byte[]>(value.Length);
                     foreach (var key in value)
                     {
-                        _keys.Add(key.ToByteArray().Reverse());
+                        var keyData = key.ToByteArray().Reverse();
```

Comment:
This one botheres me; you're adding code that is specific for a given key/algorithm to a class is supposed be key/algorithm agnostic. 

Let's discuss and find a better solution.

Reply:
Totally agree! This should be a Property in the Key-Object. But didn't dare to change 'em all.

Reply:
Can we change `public override BigInteger[] Public` to `public override byte[] Public`?
And do the conversion to whatever is expected on the wire in the key classes?

Reply:
Something like this (untested!)

```diff
From 1608712ed3e059ee001a1c989b533a97c3a7d63f Mon Sep 17 00:00:00 2001
From: Stefan Rinkes <stefan.rinkes@gmail.com>
Date: Sat, 13 Mar 2021 17:34:11 +0100
Subject: [PATCH] Better ED25519 Key fix?

---
 .../Security/Cryptography/DsaKey.cs           |  9 +++--
 .../Security/Cryptography/ED25519Key.cs       |  7 ++--
 .../Security/Cryptography/EcdsaKey.cs         | 10 ++---
 src/Renci.SshNet/Security/Cryptography/Key.cs |  2 +-
 .../Security/Cryptography/RsaKey.cs           | 11 +++---
 src/Renci.SshNet/Security/KeyHostAlgorithm.cs | 37 +++----------------
 6 files changed, 27 insertions(+), 49 deletions(-)

diff --git a/src/Renci.SshNet/Security/Cryptography/DsaKey.cs b/src/Renci.SshNet/Security/Cryptography/DsaKey.cs
index 5e9907f7..b9e76163 100644
--- a/src/Renci.SshNet/Security/Cryptography/DsaKey.cs
+++ b/src/Renci.SshNet/Security/Cryptography/DsaKey.cs
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Collections.Generic;
 using Renci.SshNet.Common;
 using Renci.SshNet.Security.Cryptography;
 
@@ -100,18 +101,18 @@ namespace Renci.SshNet.Security
         /// <value>
         /// The public.
         /// </value>
-        public override BigInteger[] Public
+        public override List<byte[]> Public
         {
             get
             {
-                return new[] { P, Q, G, Y };
+                return new List<byte[]>() { P.ToByteArray(), Q.ToByteArray(), G.ToByteArray(), Y.ToByteArray() };
             }
             set
             {
-                if (value.Length != 4)
+                if (value.Count != 4)
                     throw new InvalidOperationException("Invalid public key.");
 
-                _privateKey = value;
+                _privateKey = new[] { value[0].ToBigInteger(), value[1].ToBigInteger(), value[2].ToBigInteger(), value[3].ToBigInteger() };
             }
         }
 
diff --git a/src/Renci.SshNet/Security/Cryptography/ED25519Key.cs b/src/Renci.SshNet/Security/Cryptography/ED25519Key.cs
index 15f1cb01..79ae58f1 100644
--- a/src/Renci.SshNet/Security/Cryptography/ED25519Key.cs
+++ b/src/Renci.SshNet/Security/Cryptography/ED25519Key.cs
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Collections.Generic;
 using Renci.SshNet.Common;
 using Renci.SshNet.Security.Cryptography;
 using Renci.SshNet.Security.Chaos.NaCl;
@@ -29,15 +30,15 @@ namespace Renci.SshNet.Security
         /// <value>
         /// The public.
         /// </value>
-        public override BigInteger[] Public
+        public override List<byte[]> Public
         {
             get
             {
-                return new BigInteger[] { publicKey.ToBigInteger2() };
+                return new List<byte[]> { publicKey };
             }
             set
             {
-                publicKey = value[0].ToByteArray().Reverse().TrimLeadingZeros().Pad(Ed25519.PublicKeySizeInBytes);
+                publicKey = value[0];
             }
         }
 
diff --git a/src/Renci.SshNet/Security/Cryptography/EcdsaKey.cs b/src/Renci.SshNet/Security/Cryptography/EcdsaKey.cs
index 244f316e..a06a849a 100644
--- a/src/Renci.SshNet/Security/Cryptography/EcdsaKey.cs
+++ b/src/Renci.SshNet/Security/Cryptography/EcdsaKey.cs
@@ -1,5 +1,6 @@
 ﻿#if FEATURE_ECDSA
 using System;
+using System.Collections.Generic;
 using System.IO;
 using System.Text;
 using System.Runtime.InteropServices;
@@ -137,7 +138,7 @@ namespace Renci.SshNet.Security
         /// <value>
         /// The public.
         /// </value>
-        public override BigInteger[] Public
+        public override List<byte[]> Public
         {
             get
             {
@@ -200,15 +201,14 @@ namespace Renci.SshNet.Security
                 Buffer.BlockCopy(qy, 0, q, qx.Length + 1, qy.Length);
 
                 // returns Curve-Name and x/y as ECPoint
-                return new[] { new BigInteger(curve.Reverse()), new BigInteger(q.Reverse()) };
+                return  new List<byte[]>() { curve, q };
             }
             set
             {
-                var curve_s = Encoding.ASCII.GetString(value[0].ToByteArray().Reverse());
+                var curve_s = Encoding.ASCII.GetString(value[0]);
                 string curve_oid = GetCurveOid(curve_s);
 
-                var publickey = value[1].ToByteArray().Reverse();
-                Import(curve_oid, publickey, null);
+                Import(curve_oid, value[1], null);
             }
         }
 
diff --git a/src/Renci.SshNet/Security/Cryptography/Key.cs b/src/Renci.SshNet/Security/Cryptography/Key.cs
index 61c5150c..412f3a4a 100644
--- a/src/Renci.SshNet/Security/Cryptography/Key.cs
+++ b/src/Renci.SshNet/Security/Cryptography/Key.cs
@@ -26,7 +26,7 @@ namespace Renci.SshNet.Security
         /// <value>
         /// The public.
         /// </value>
-        public abstract BigInteger[] Public { get; set; }
+        public abstract List<byte[]> Public { get; set; }
 
         /// <summary>
         /// Gets the length of the key.
diff --git a/src/Renci.SshNet/Security/Cryptography/RsaKey.cs b/src/Renci.SshNet/Security/Cryptography/RsaKey.cs
index ba5b1746..e025c2f7 100644
--- a/src/Renci.SshNet/Security/Cryptography/RsaKey.cs
+++ b/src/Renci.SshNet/Security/Cryptography/RsaKey.cs
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Collections.Generic;
 using Renci.SshNet.Common;
 using Renci.SshNet.Security.Cryptography;
 
@@ -154,18 +155,18 @@ namespace Renci.SshNet.Security
         /// <value>
         /// The public.
         /// </value>
-        public override BigInteger[] Public
+        public override List<byte[]> Public
         {
             get
             {
-                return new[] { Exponent, Modulus };
+                return new List<byte[]> { Exponent.ToByteArray(), Modulus.ToByteArray() };
             }
             set
             {
-                if (value.Length != 2)
-                    throw new InvalidOperationException("Invalid private key.");
+                if (value.Count != 2)
+                    throw new InvalidOperationException("Invalid public key.");
 
-                _privateKey = new[] { value[1], value[0] };
+                _privateKey = new[] { value[1].ToBigInteger(), value[0].ToBigInteger() };
             }
         }
 
diff --git a/src/Renci.SshNet/Security/KeyHostAlgorithm.cs b/src/Renci.SshNet/Security/KeyHostAlgorithm.cs
index 11717197..9986fd03 100644
--- a/src/Renci.SshNet/Security/KeyHostAlgorithm.cs
+++ b/src/Renci.SshNet/Security/KeyHostAlgorithm.cs
@@ -83,33 +83,8 @@ namespace Renci.SshNet.Security
         private class SshKeyData : SshData
         {
             private byte[] _name;
-            private List<byte[]> _keys;
 
-            public BigInteger[] Keys
-            {
-                get
-                {
-                    var keys = new BigInteger[_keys.Count];
-                    for (var i = 0; i < _keys.Count; i++)
-                    {
-                        var key = _keys[i];
-                        keys[i] = key.ToBigInteger2();
-                    }
-                    return keys;
-                }
-                private set
-                {
-                    _keys = new List<byte[]>(value.Length);
-                    foreach (var key in value)
-                    {
-                        var keyData = key.ToByteArray().Reverse();
-                        if (Name == "ssh-ed25519")
-                            keyData = keyData.TrimLeadingZeros().Pad(Ed25519.PublicKeySizeInBytes);
-
-                        _keys.Add(keyData);
-                    }
-                }
-            }
+            public List<byte[]> Keys { get; private set; }
 
             private string Name
             {
@@ -124,7 +99,7 @@ namespace Renci.SshNet.Security
                     var capacity = base.BufferCapacity;
                     capacity += 4; // Name length
                     capacity += _name.Length; // Name
-                    foreach (var key in _keys)
+                    foreach (var key in Keys)
                     {
                         capacity += 4; // Key length
                         capacity += key.Length; // Key
@@ -137,7 +112,7 @@ namespace Renci.SshNet.Security
             {
             }
 
-            public SshKeyData(string name, params BigInteger[] keys)
+            public SshKeyData(string name, List<byte[]> keys)
             {
                 Name = name;
                 Keys = keys;
@@ -146,11 +121,11 @@ namespace Renci.SshNet.Security
             protected override void LoadData()
             {
                 _name = ReadBinary();
-                _keys = new List<byte[]>();
+                Keys = new List<byte[]>();
 
                 while (!IsEndOfData)
                 {
-                    _keys.Add(ReadBinary());
+                    Keys.Add(ReadBinary());
                 }
             }
 
@@ -158,7 +133,7 @@ namespace Renci.SshNet.Security
             {
                 WriteBinaryString(_name);
 
-                foreach (var key in _keys)
+                foreach (var key in Keys)
                 {
                     WriteBinaryString(key);
                 }
-- 
2.30.0.windows.2
```

Reply:
@darinkes: Personally, I don't like having **List<byte[]>** as type for the *Public* property.
That doesn't mean we should abandon the idea though.
Another option would be for each algorithm to have its property factory to construct a **KeyHostAlgorithm**.

```cs
/// <summary>
/// Produces a given <see cref="KeyHostAlgorithm"/>.
/// </summary>
public interface IHostKeyAlgorithmFactory
{
    /// <summary>
    /// Gets the name of the host key algorithm that this factory can produce <see cref="KeyHostAlgorithm"/>
    /// instances for. 
    /// </summary>
    /// <value>
    /// The name of the host key algorithm that this factory can produce <see cref="KeyHostAlgorithm"/>
    /// instances for. 
    /// </value>
    string Name { get; }

    /// <summary>
    /// Create a <see cref="KeyHostAlgorithm"/> from the key data that the host provided.
    /// </summary>
    /// <param name="hostKeyData">The host key data.</param>
    /// <returns>
    /// A <see cref="KeyHostAlgorithm"/> created from the key data that the host provided.
    /// </returns>
    KeyHostAlgorithm Create(byte[] hostKeyData);
}

public class ED25519HostKeyAlgorithmFactory : IHostKeyAlgorithmFactory
{
    public string Name
    {
        get
        {
            return "ssh-ed25519";
        }
    }

    public KeyHostAlgorithm Create(byte[] hostKeyData)
    {
        var sshKey = new SshKeyData();
        sshKey.Load(hostKeyData);
        ...
    }
}

...
```

We'd have to make **KeyHostAlgorithm.SshKeyData** internal, and perhaps move it outside of **KeyHostAlgorithm**.
The **Keys** property of that class should effectively be changed to **List<byte[]>**. This bothers me less since its an internal class.

I also wonder if we shouldn't create public key classes instead of "abusing" the private key classes, and split the **KeyHostAlgorithm** class into a class for authentication and one for verifying the host key.

The **Name** property on **IHostKeyAlgorithmFactory** would - at one point - be used as key for the **HostKeyAlgorithms** dictionary on **ConnectionInfo**. I'd like to evolve to a fluent builder syntax, but that's for another time.

The factory pattern also allows us to remove the default constructors of our **Key** classes.

Would it possible for you to discuss both propositions via Skype or WhatsApp?

Reply:
The proposed `List<byte[]>` also includes the idea to ease up Key Export.
See: https://github.com/darinkes/SshNet.Keygen/blob/main/SshNet.Keygen/Extensions/KeyExtension.cs#L254

Instead of calling different stuff for each KeyClass, call Key.Public and encode each byte-array.
OpenSSH-Key format is identical to the wire-protocol for Key-Exchange.

Whats the issue with `List<byte[]>`? IMHO its a clean and easy to understand representation of the SSH-Protocol.

Reply:
**List<byte[]>** is just a very loose construct, by itself its not that much of an issue. It's also that with the current implementation (and this has nothing to do with this PR):
* all key classes need to have a public constructor hereby allowing for key instances that are not fully initialized.
* all key classes are sometimes used to represent a private key and other times to represent a public key; in case of the latter some properties will throw or return bogus information.

If we are to introduce a breaking change in our API, I'd just like to consider more options.

Reply:
Hm, then my first attempt is the less breaking one and it can/should be improved later?